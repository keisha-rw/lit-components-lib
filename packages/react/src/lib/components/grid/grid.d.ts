import * as React from 'react';
export declare const PdsGrid: React.FunctionComponent<Partial<{
    variant: "default" | "2up" | "3up" | "1-3up" | "1-2-4up" | "side-by-side";
    gap: "default" | "sm" | "lg" | "none";
    break: "default" | "faster" | "slower";
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Partial<{}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
