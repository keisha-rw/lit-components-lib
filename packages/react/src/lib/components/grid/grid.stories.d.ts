import * as React from 'react';
declare const _default: {
    title: string;
    component: React.FunctionComponent<Partial<{
        variant: "default" | "2up" | "3up" | "1-3up" | "1-2-4up" | "side-by-side";
        gap: "default" | "sm" | "lg" | "none";
        break: "default" | "faster" | "slower";
    }> & {
        ref?: React.Ref<any> | undefined;
    } & {
        style?: React.CSSProperties | undefined;
    } & Partial<{}> & {
        id?: string | undefined;
        slot?: string | undefined;
        className?: string | undefined;
        children?: React.ReactNode;
    }>;
    tags: string[];
};
export default _default;
export declare const Default: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0fc72a6d").R, Partial<{
    variant: "default" | "2up" | "3up" | "1-3up" | "1-2-4up" | "side-by-side";
    gap: "default" | "sm" | "lg" | "none";
    break: "default" | "faster" | "slower";
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Partial<{}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
export declare const SideBySide: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0fc72a6d").R, Partial<{
    variant: "default" | "2up" | "3up" | "1-3up" | "1-2-4up" | "side-by-side";
    gap: "default" | "sm" | "lg" | "none";
    break: "default" | "faster" | "slower";
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Partial<{}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
export declare const TwoUp: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0fc72a6d").R, Partial<{
    variant: "default" | "2up" | "3up" | "1-3up" | "1-2-4up" | "side-by-side";
    gap: "default" | "sm" | "lg" | "none";
    break: "default" | "faster" | "slower";
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Partial<{}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
export declare const ThreeUp: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0fc72a6d").R, Partial<{
    variant: "default" | "2up" | "3up" | "1-3up" | "1-2-4up" | "side-by-side";
    gap: "default" | "sm" | "lg" | "none";
    break: "default" | "faster" | "slower";
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Partial<{}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
export declare const ThreeUpBreakFaster: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0fc72a6d").R, Partial<{
    variant: "default" | "2up" | "3up" | "1-3up" | "1-2-4up" | "side-by-side";
    gap: "default" | "sm" | "lg" | "none";
    break: "default" | "faster" | "slower";
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Partial<{}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
export declare const ThreeUpBreakSlower: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0fc72a6d").R, Partial<{
    variant: "default" | "2up" | "3up" | "1-3up" | "1-2-4up" | "side-by-side";
    gap: "default" | "sm" | "lg" | "none";
    break: "default" | "faster" | "slower";
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Partial<{}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
export declare const OneToThreeUp: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0fc72a6d").R, Partial<{
    variant: "default" | "2up" | "3up" | "1-3up" | "1-2-4up" | "side-by-side";
    gap: "default" | "sm" | "lg" | "none";
    break: "default" | "faster" | "slower";
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Partial<{}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
export declare const OneToThreeUpBreakFaster: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0fc72a6d").R, Partial<{
    variant: "default" | "2up" | "3up" | "1-3up" | "1-2-4up" | "side-by-side";
    gap: "default" | "sm" | "lg" | "none";
    break: "default" | "faster" | "slower";
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Partial<{}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
export declare const OneToThreeUpBreakSlower: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0fc72a6d").R, Partial<{
    variant: "default" | "2up" | "3up" | "1-3up" | "1-2-4up" | "side-by-side";
    gap: "default" | "sm" | "lg" | "none";
    break: "default" | "faster" | "slower";
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Partial<{}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
export declare const OneToTwoToFourUp: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0fc72a6d").R, Partial<{
    variant: "default" | "2up" | "3up" | "1-3up" | "1-2-4up" | "side-by-side";
    gap: "default" | "sm" | "lg" | "none";
    break: "default" | "faster" | "slower";
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Partial<{}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
export declare const OneToTwoToFourUpBreakFaster: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0fc72a6d").R, Partial<{
    variant: "default" | "2up" | "3up" | "1-3up" | "1-2-4up" | "side-by-side";
    gap: "default" | "sm" | "lg" | "none";
    break: "default" | "faster" | "slower";
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Partial<{}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
export declare const OneToTwoToFourUpBreakSlower: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0fc72a6d").R, Partial<{
    variant: "default" | "2up" | "3up" | "1-3up" | "1-2-4up" | "side-by-side";
    gap: "default" | "sm" | "lg" | "none";
    break: "default" | "faster" | "slower";
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Partial<{}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
