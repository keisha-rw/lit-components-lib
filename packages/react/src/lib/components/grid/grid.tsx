import { PdsGrid as PdsGridElement } from '@keisha/design-system';
import * as React from 'react';
import { PdsGridProps } from '../../built-component-props';
import { createPdsReactComponent } from '../../create-pds-react-component';

export const PdsGrid = createPdsReactComponent<PdsGridProps>()({
  tagName: 'pds-grid',
  elementClass: PdsGridElement,
  react: React,
});
