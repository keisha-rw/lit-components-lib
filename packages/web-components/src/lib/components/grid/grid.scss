/*------------------------------------*\
 #GRID
\*------------------------------------*/

:host {
  display: block;
}

/**
 * 1) Grid layout for items like cards, etc.
 */
.pds-l-grid {
  --gap-size: calc(var(--pds-base-unit) * 3);

  display: flex;
  flex-flow: column;
  gap: var(--gap-size);
}

.pds-l-grid--gap-none {
  --gap-size: 0;

  gap: var(--gap-size);
}

.pds-l-grid--gap-sm {
  --gap-size: var(--pds-base-unit);

  gap: var(--gap-size);
}

.pds-l-grid--gap-lg {
  --gap-size: calc(var(--pds-base-unit) * 4);

  gap: var(--gap-size);
}

/**
 * 2up grid
 * 1) Stacked items on small screens to 2 items per row on medium/large screens
 */
.pds-l-grid--2up {
  @media all and (min-width: $pds-bp-lg) {
    flex-flow: row wrap;
  }
}

/**
 * 2up grid break faster
 * Break 1 breakpoint smaller than default
 */
.pds-l-grid--2up.pds-l-grid--break-faster {
  @media all and (min-width: $pds-bp-md) {
    flex-flow: row wrap;
  }
}

/**
 * 1-3up grid
 * 1) Stacked items on small screens to 3up items per row on medium/large screens
 */
.pds-l-grid--1-3up {
  @media all and (min-width: $pds-bp-md) {
    flex-flow: row wrap;
  }
}

/**
 * 1-3up grid break faster
 * Break 1 breakpoint smaller than default
 */
.pds-l-grid--1-3up.pds-l-grid--break-faster {
  @media all and (min-width: $pds-bp-sm) {
    flex-flow: row wrap;
  }
}

/**
 * 1-2-4up grid
 * 1) Stacked items on small screens to 2 items per row on medium/large screens
 */
.pds-l-grid--1-2-4up {
  @media all and (min-width: $pds-bp-md) {
    flex-flow: row wrap;
  }
}

/**
 * 1-2-4up grid break faster
 * Break 1 breakpoint smaller than default
 */
.pds-l-grid--1-2-4up.pds-l-grid--break-faster {
  @media all and (min-width: $pds-bp-sm) {
    flex-flow: row wrap;
  }
}

/**
 * side-by-side grid
 */
.pds-l-grid--side-by-side {
  flex-flow: row wrap;
}

/**
 * 3up grid
 * 1) Stacked items on small screens to 2/3up items per row on medium/large screens
 */
.pds-l-grid--3up {
  @media all and (min-width: $pds-bp-md) {
    flex-flow: row wrap;
  }
}

.pds-l-grid--3up.pds-l-grid--break-faster {
  @media all and (min-width: $pds-bp-sm) {
    flex-flow: row wrap;
  }
}

::slotted(pds-grid-item) {
  display: block;
  width: 100%;

  /**
  * Slotted grid item within 2up Grid
  */
  .pds-l-grid--2up > & {
    @media all and (min-width: $pds-bp-lg) {
      width: calc((100% - var(--gap-size)) / 2);
    }
  }

  /**
  * Slotted grid item within 1-3up Grid
  */
  .pds-l-grid--1-3up > & {
    @media all and (min-width: $pds-bp-md) {
      --gap-total: calc(var(--gap-size) * 2);

      width: calc((100% - var(--gap-total)) / 3);
    }
  }

  /**
  * Slotted grid item within 1-2-4up Grid
  */
  .pds-l-grid--1-2-4up > & {
    @media all and (min-width: $pds-bp-md) {
      width: calc((100% - var(--gap-size)) / 2);
    }

    @media all and (min-width: $pds-bp-lg) {
      --gap-total: calc(var(--gap-size) * 3);

      width: calc((100% - var(--gap-total)) / 4);
    }
  }

  /**
  * Slotted grid item within side-by-side Grid
  */
  .pds-l-grid--side-by-side > & {
    width: calc((100% - var(--gap-size)) / 2);
  }

  /**
  * Slotted grid item within 3up Grid
  */
  .pds-l-grid--3up > & {
    @media all and (min-width: $pds-bp-md) {
      width: calc((100% - var(--gap-size)) / 2);
    }

    @media all and (min-width: $pds-bp-lg) {
      --gap-total: calc(var(--gap-size) * 2);

      width: calc((100% - var(--gap-total)) / 3);
    }
  }

  /**
  * Slotted grid item within grid with no gaps in between
  */
  .pds-l-grid--gap-none > & {
    --gap-size: 0px;
  }

  /**
  * Slotted grid item within grid with small gap between items than default
  */
  .pds-l-grid--gap-sm > & {
    --gap-size: var(--pds-base-unit);
  }

  /**
  * Slotted grid item within grid with large gap between items than default
  */
  .pds-l-grid--gap-lg > & {
    --gap-size: calc(var(--pds-base-unit) * 4);
  }

  /**
  * Slotted grid item within 2up Grid that breaks faster than default
  */
  .pds-l-grid--2up.pds-l-grid--break-faster > & {
    @media all and (min-width: $pds-bp-md) {
      width: calc((100% - var(--gap-size)) / 2);
    }
  }

  /**
  * Slotted grid item within 2up Grid that breaks slower than default
  */
  .pds-l-grid--2up.pds-l-grid--break-slower > & {
    @media all and (min-width: $pds-bp-lg) {
      width: 100%;
    }

    @media all and (min-width: $pds-bp-xl) {
      width: calc((100% - var(--gap-size)) / 2);
    }
  }

  /**
  * Slotted grid item within 1-2-4up Grid that breaks faster than default
  */
  .pds-l-grid--1-2-4up.pds-l-grid--break-faster > & {
    @media all and (min-width: $pds-bp-sm) {
      width: calc((100% - var(--gap-size)) / 2);
    }

    @media all and (min-width: $pds-bp-md) {
      --gap-total: calc(var(--gap-size) * 3);

      width: calc((100% - var(--gap-total)) / 4);
    }
  }

  /**
  * Slotted grid item within 1-2-4up Grid that breaks faster than default
  */
  .pds-l-grid--1-2-4up.pds-l-grid--break-slower > & {
    @media all and (min-width: $pds-bp-md) {
      width: 100%;
    }

    @media all and (min-width: $pds-bp-lg) {
      width: calc((100% - var(--gap-size)) / 2);
    }

    @media all and (min-width: $pds-bp-xl) {
      --gap-total: calc(var(--gap-size) * 3);

      width: calc((100% - var(--gap-total)) / 4);
    }
  }

  /**
  * Slotted grid item within 3up Grid that breaks faster than default
  */
  .pds-l-grid--1-3up.pds-l-grid--break-faster > & {
    @media all and (min-width: $pds-bp-sm) {
      --gap-total: calc(var(--gap-size) * 2);

      width: calc((100% - var(--gap-total)) / 3);
    }
  }

  /**
  * Slotted grid item within 3up Grid that breaks faster than default
  */
  .pds-l-grid--1-3up.pds-l-grid--break-slower > & {
    @media all and (min-width: $pds-bp-md) {
      width: 100%;
    }

    @media all and (min-width: $pds-bp-lg) {
      --gap-total: calc(var(--gap-size) * 2);

      width: calc((100% - var(--gap-total)) / 3);
    }
  }

  /**
  * Slotted grid item within 3up Grid that breaks faster than default
  */
  .pds-l-grid--3up.pds-l-grid--break-faster > & {
    @media all and (min-width: $pds-bp-sm) {
      width: calc((100% - var(--gap-size)) / 2);
    }

    @media all and (min-width: $pds-bp-md) {
      --gap-total: calc(var(--gap-size) * 2);

      width: calc((100% - var(--gap-total)) / 3);
    }
  }

  /**
  * Slotted grid item within 3up Grid that breaks slower than default
  */
  .pds-l-grid--3up.pds-l-grid--break-slower > & {
    @media all and (min-width: $pds-bp-lg) {
      width: calc((100% - var(--gap-size)) / 2);
    }

    @media all and (min-width: $pds-bp-xl) {
      --gap-total: calc(var(--gap-size) * 2);

      width: calc((100% - var(--gap-total)) / 3);
    }
  }
}
