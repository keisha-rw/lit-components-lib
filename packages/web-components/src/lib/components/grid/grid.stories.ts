import { Meta, StoryObj } from '@storybook/web-components';
import { html, nothing } from 'lit';
import './grid';
import '../grid-item/grid-item';
import '../../../../.storybook/components/placeholder-element/placeholder-element';

export default {
  title: 'Layout/Grid',
  component: 'pds-grid',
  tags: ['autodocs'],
  render: (args) =>
    html`<pds-grid
      variant="${args.variant || nothing}"
      break="${args.break || nothing}"
      ><pds-grid-item>
        <placeholder-element>Item 1</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 2</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 3</placeholder-element>
      </pds-grid-item></pds-grid
    >`,
} as Meta;

// span element is needed to wrap plain text slots https://github.com/storybookjs/storybook/issues/15963

export const Default: StoryObj = {
  name: 'Default',
  args: {},
};

export const SideBySide: StoryObj = {
  name: 'Side by side',
  args: {
    variant: 'side-by-side',
  },
  render: (args) =>
    html`<pds-grid variant="${args.variant}">
      <pds-grid-item>
        <placeholder-element>Item 1</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 2</placeholder-element>
      </pds-grid-item>
    </pds-grid>`,
};

export const TwoUp: StoryObj = {
  name: '2up',
  args: {
    variant: '2up',
  },
  render: (args) =>
    html`<pds-grid variant="${args.variant}">
      <pds-grid-item>
        <placeholder-element>Item 1</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 2</placeholder-element>
      </pds-grid-item>
    </pds-grid>`,
};

export const ThreeUp: StoryObj = {
  name: '3up',
  args: {
    variant: '3up',
  },
  render: (args) =>
    html`<pds-grid variant="${args.variant}">
      <pds-grid-item>
        <placeholder-element>Item 1</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 2</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 3</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 4</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 5</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 6</placeholder-element>
      </pds-grid-item>
    </pds-grid>`,
};

export const ThreeUpBreakFaster: StoryObj = {
  name: '3up break faster',
  args: {
    variant: '3up',
    break: 'faster',
  },
  render: (args) =>
    html`<pds-grid variant="${args.variant}" break="${args.break}">
      <pds-grid-item>
        <placeholder-element>Item 1</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 2</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 3</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 4</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 5</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 6</placeholder-element>
      </pds-grid-item>
    </pds-grid>`,
};

export const ThreeUpBreakSlower: StoryObj = {
  name: '3up break slower',
  args: {
    variant: '3up',
    break: 'slower',
  },
  render: (args) =>
    html`<pds-grid variant="${args.variant}" break="${args.break}">
      <pds-grid-item>
        <placeholder-element>Item 1</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 2</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 3</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 4</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 5</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 6</placeholder-element>
      </pds-grid-item>
    </pds-grid>`,
};

export const OneToThreeUp: StoryObj = {
  name: '1-3up',
  args: {
    variant: '1-3up',
  },
  render: (args) =>
    html`<pds-grid variant="${args.variant}">
      <pds-grid-item>
        <placeholder-element>Item 1</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 2</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 3</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 4</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 5</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 6</placeholder-element>
      </pds-grid-item>
    </pds-grid>`,
};

export const OneToThreeUpBreakFaster: StoryObj = {
  name: '1-3up break faster',
  args: {
    variant: '1-3up',
    break: 'faster',
  },
  render: (args) =>
    html`<pds-grid variant="${args.variant}" break="${args.break}">
      <pds-grid-item>
        <placeholder-element>Item 1</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 2</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 3</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 4</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 5</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 6</placeholder-element>
      </pds-grid-item>
    </pds-grid>`,
};

export const OneToThreeUpBreakSlower: StoryObj = {
  name: '1-3up break slower',
  args: {
    variant: '1-3up',
    break: 'slower',
  },
  render: (args) =>
    html`<pds-grid variant="${args.variant}" break="${args.break}">
      <pds-grid-item>
        <placeholder-element>Item 1</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 2</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 3</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 4</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 5</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 6</placeholder-element>
      </pds-grid-item>
    </pds-grid>`,
};

export const OneToTwoToFourUp: StoryObj = {
  name: '1-2-4up',
  args: {
    variant: '1-2-4up',
  },
  render: (args) =>
    html`<pds-grid variant="${args.variant}">
      <pds-grid-item>
        <placeholder-element>Item 1</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 2</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 3</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 4</placeholder-element>
      </pds-grid-item>
    </pds-grid>`,
};
export const OneToTwoToFourUpBreakFaster: StoryObj = {
  name: '1-2-4up break faster',
  args: {
    variant: '1-2-4up',
    break: 'faster',
  },
  render: (args) =>
    html`<pds-grid variant="${args.variant}" break="${args.break}">
      <pds-grid-item>
        <placeholder-element>Item 1</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 2</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 3</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 4</placeholder-element>
      </pds-grid-item>
    </pds-grid>`,
};
export const OneToTwoToFourUpBreakSlower: StoryObj = {
  name: '1-2-4up break slower',
  args: {
    variant: '1-2-4up',
    break: 'slower',
  },
  render: (args) =>
    html`<pds-grid variant="${args.variant}" break="${args.break}">
      <pds-grid-item>
        <placeholder-element>Item 1</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 2</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 3</placeholder-element>
      </pds-grid-item>
      <pds-grid-item>
        <placeholder-element>Item 4</placeholder-element>
      </pds-grid-item>
    </pds-grid>`,
};
