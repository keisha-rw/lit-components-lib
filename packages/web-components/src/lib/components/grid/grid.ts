import { html } from 'lit';
import { property } from 'lit/decorators.js';
import { classMap } from 'lit/directives/class-map.js';
import { pdsCustomElement as customElement } from '../../decorators/pds-custom-element';
import { PdsElement } from '../PdsElement';
import '../grid-item/grid-item';
import styles from './grid.scss?inline';

/**
 * @summary This component provides some helpful layout configurations
 *
 * @slot default Contains grid items within the layout configuration, restricted to pds-grid-item elements
 */
@customElement('pds-grid', {
  category: 'layout',
  type: 'component',
  state: 'stable',
  styles,
})
export class PdsGrid extends PdsElement {
  /**
   * - **default** renders the grid item stacked with a gap in between
   * - **2up** yields a grid whose grid items are stacked on small screens but display side-by-side when enough screen real estate is available to do so
   * - **3up** yields a grid whose grid items are stacked on small screens, transforms to a 2-across pattern, and ultimately transforms to a 3-across pattern
   * - **1-2-4up** yields a grid whose grid items are stacked on small screens, transforms to a 2-across pattern, and ultimately transforms to a 4-across pattern
   * - **1-3up** yields a grid whose grid items are stacked on small screens transforms to a 3-across pattern on larger screens
   * - **side-by-side** yields a grid whose grid items don't necessarily ever need to stack and displayed side-by-side for all the viewports
   */
  @property()
  variant: '2up' | '3up' | '1-3up' | '1-2-4up' | 'side-by-side' | 'default' =
    'default';

  /**
   * - **default** grid items are spaced with a gap
   * - **none** grid items are spaced with no gap between items
   * - **sm** grid items are spaced with smaller gap than default
   * - **lg** grid items are spaced with large gap than default
   */
  @property()
  gap: 'none' | 'sm' | 'lg' | 'default' = 'default';

  /**
   * - **default** stacks at default viewport
   * - **faster** stacks at a smaller viewport than the default
   * - **slower** stacks at a larger viewport than the default
   */
  @property()
  break: 'faster' | 'slower' | 'default' = 'default';

  render() {
    const classes = {
      'pds-l-grid': true,
      [`pds-l-grid--${this.variant}`]: !!this.variant,
      [`pds-l-grid--gap-${this.gap}`]: !!this.gap,
      [`pds-l-grid--break-${this.break}`]: this.break,
    };

    return html`<div class=${classMap(classes)}>
      <slot
        allowed-elements="pds-grid-item"
        @slotchange=${this.handleSlotValidation}
      ></slot>
    </div>`;
  }
}
