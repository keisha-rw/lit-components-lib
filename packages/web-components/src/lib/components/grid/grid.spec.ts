import * as path from 'path';
import initStoryshots from '@storybook/addon-storyshots';
import { assert, fixture } from '@open-wc/testing';
import { PdsGrid } from './grid';
import createSnapshots from '../../../../test/utils/snapshots';

initStoryshots({
  storyKindRegex: /Grid/,
  framework: 'web-components',
  configPath: path.join(__dirname, '../../../../.storybook'),
  asyncJest: true,
  test: createSnapshots,
});

describe('Grid unit tests', () => {
  let element: Element;

  beforeEach(async () => {
    element = await fixture('<pds-grid/>');
  });

  it('is an instance of PdsGrid', () => {
    expect(element).toBeInstanceOf(PdsGrid);
  });

  it('is accessible', async () => {
    await assert.isAccessible(element);
  });

  it('populates the default slot correctly with pds-grid-item', async () => {
    const consoleError = jest
      .spyOn(console, 'error')
      .mockImplementation(() => {});
    element = await fixture(
      '<pds-grid><pds-grid-item><placeholder-element>Item 1</placeholder-element></pds-grid-item></pds-grid>',
    );
    expect(consoleError).not.toHaveBeenCalled();
    consoleError.mockRestore();
  });

  it('does not populate the default slot correctly when the provided element is not a pds-grid-item', async () => {
    const consoleError = jest
      .spyOn(console, 'error')
      .mockImplementation(() => {});
    element = await fixture(
      '<pds-grid><div><placeholder-element>Item 1</placeholder-element></div></pds-grid>',
    );
    expect(element).toMatchSnapshot();
    expect(consoleError).toHaveBeenCalledTimes(1);
    consoleError.mockRestore();
  });
});
