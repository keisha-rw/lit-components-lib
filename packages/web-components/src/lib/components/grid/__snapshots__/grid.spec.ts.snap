// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid unit tests does not populate the default slot correctly when the provided element is not a pds-grid-item 1`] = `<pds-grid />`;
